dash/dash.py

```diff
@@ -1035,12 +1111,111 @@ def enable_dev_tools(self,
             default=debug,
             is_bool=True
         )
+        self._dev_tools['hot_reload'] = _configs.get_config(
+            'hot_reload', dev_tools_hot_reload, env,
+            default=debug,
+            is_bool=True
+        )
+        self._dev_tools['hot_reload_interval'] = int(_configs.get_config(
+            'hot_reload_interval', dev_tools_hot_reload_interval, env,
+            default=3000
+        ))
+        self._dev_tools['hot_reload_watch_interval'] = float(
+            _configs.get_config(
+                'hot_reload_watch_interval',
+                dev_tools_hot_reload_watch_interval,
+                env,
+                default=0.5
+            )
+        )
+        self._dev_tools['hot_reload_max_retry'] = int(
+            _configs.get_config(
+                'hot_reload_max_retry',
+                dev_tools_hot_reload_max_retry,
+                env,
+                default=8
+            )
+        )
+        self._dev_tools['silence_routes_logging'] = _configs.get_config(
+            'silence_routes_logging', dev_tools_silence_routes_logging, env,
+            default=debug,
+            is_bool=True,
+        )
+
+        if self._dev_tools.silence_routes_logging:
+            logging.getLogger('werkzeug').setLevel(logging.ERROR)
+            self.logger.setLevel(logging.INFO)
+
+        if self._dev_tools.hot_reload:
+            self._reload_hash = _generate_hash()
+            self._watch_thread = threading.Thread(
+                target=lambda: _watch.watch(
+                    [self._assets_folder],
+                    self._on_assets_change,
+                    sleep_time=self._dev_tools.hot_reload_watch_interval)
+            )
+            self._watch_thread.daemon = True
+            self._watch_thread.start()
+
+        if debug and self._dev_tools.serve_dev_bundles:
+            # Dev bundles only works locally.
+            self.scripts.config.serve_locally = True
+
         return debug
 
+    # noinspection PyProtectedMember
+    def _on_assets_change(self, filename, modified, deleted):
+        self._lock.acquire()
+        self._hard_reload = True
+        self._reload_hash = _generate_hash()
+
+        asset_path = filename.replace(self._assets_folder, '')\
+            .replace('\\', '/').lstrip('/')
+
+        self._changed_assets.append({
+            'url': self.get_asset_url(asset_path),
+            'modified': int(modified),
+            'is_css': filename.endswith('css')
+        })
+
+        if filename not in self._assets_files and not deleted:
+            res = self._add_assets_resource(asset_path, filename)
+            if filename.endswith('js'):
+                self.scripts.append_script(res)
+            elif filename.endswith('css'):
+                self.css.append_css(res)
+
+        if deleted:
+            if filename in self._assets_files:
+                self._assets_files.remove(filename)
+
+            def delete_resource(resources):
```

Comment:
Can this be a bit more pythonic? It was a bit hard to read through the first time.

>>> to_delete = next(filter(lambda r: r.get('asset_path') == asset_path, resources), None)
>>> if to_delete:
...     resources.remove(to_delete)

Reply:
I must admit it looks weird, but `filter` is not pythonic according to Guido. It would also iterate over all the array while the `for` loop stops once found.